### DISCUSSION TITLE
Launching debugger in vscode for fpm projects

### DISCUSSION THREAD

**Discussion by bitli** [fortran-lang/fpm#594] [General]
Hello,
I have successfully used _vscode_ tasks and launcher to compile, launch and debug main programs from _vscode_ and fpm. The difficulty was to find the location of the program to launch created in a generated directory as `build/gfortran_2A42023B310FA28D/bin`. The workaround was to create a `work` directory somewhere and use `fpm install --prefix .../work` instead of `fpm build`, so the binaries are at a predictable location. Then the `launch.json` can specify `"program": "work/main"` for the debugger and all works fine.
However when developing a library module I must test using programs in the _test_ or _example_ directories and I did not find a way to automatically copy the files to the work directory. May be a shell script could be called after the build task, but this is not very convenient especially in a multi-architecture project. I tried to use the command `fpm test --runner gdb`, but then the debug interface of _vscode_ try to connect to fpm, not to the started gdb (or other strange things happens). 
Any idea of what could be done ? 
Ideally I would like to install the examples or test in a specific location, just like `fpm install` or build the project in a fixed directory or get the location of the build directory (this is less convenient).

**Comment by zoziha** [comment]
I am also using vscode (I think vscode and fpm work well together). When debugging, it is troublesome to find the correct executable program path. Maybe we can find a way.

I usually look up the executable program generation location patiently based on the running information of fpm (it seems that the location information contains a hash value), and then manually paste it into `launch.json`.

Digression (not important at this stage): I see that vscode has official plug-ins for languages such as julia and python. I wonder if the fortran community specifically wants to promote Fortran programming in vscode?

```sh
fpm test tester --runner ls
```

**Reply by bitli** [reply]
I used the same trick initially
But finally I use the workaround (works on Linux and for a specific executable):
I created a script as follow:
```
#!/bin/sh
# Find and isntall the example executables to ../release/bin
mkdir -p release/bin
rm -f release/bin/*
# Use the most recent output
GFORTRAN_DIR=`ls build -t | grep -m 1 gfortran_`
cp build/${GFORTRAN_DIR}/example/* release/bin
echo "Copied executables: " `ls release/bin`
```
in tasks.json, I use a new defautl build task that depends on the classic fpmbuild task:
```
       {
            "label": "prepare",
            "type": "shell",
            "command": "${workspaceFolder}/devenvs/vscode/install-examples.sh",
            "args": [
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "dependsOn": ["fpmbuild"],
```
So the launch.json can have a fixed program to launch:
```
            ...
            "program": "${workspaceRoot}/release/bin/LoadWaveInfo",
            ...
            "preLaunchTask": "prepare",
            ...
```
This could be enhanced, for example to only copy the program named like the active windows if I have multiple examples and so on. But this is very environment, project and os specific, unfortunately.

**Comment by zoziha** [comment]
Hello, @bitli :

I am also a user of `vs code`. `fpm` runs in the built-in terminal of `vs code` and I write code using `vs code`. I think this experience is good, but as you said, configuring the `launch.json` file takes some time.
Based on interest, I quickly and simply wrote a small program `vdb` using Fortran to do this, and let it configure the `launch.json` file according to the arguments passed from `fpm`.
The effect is as follows:
![start.gif](https://i.loli.net/2021/11/14/8mSWJc5iBC7KA3I.gif)
The currently supported commands are:
```sh
fpm [options] <test_name> --runner vdb
fpm [options] <test_name> --runner vdb - ARGS
vdb clean # Remove and backup JSON::launch.json
```

**Note**: Because some file processing commands are used, such as `mkdir` and `mv`, I cannot guarantee that it will work under other systems, but it will work under `windows 10 + msys2-gfortran + pwsh`.

### Link

- [zoziha/vdb](https://github.com/zoziha/vdb)
